### PR TITLE
Add Sphincs+ To Signer Interface

### PR DESCRIPTION
**Pull request by rugo**
Fixes: 
Updates: #169

### Description of the changes being introduced by the pull request:
The PRs #160 and #169  tried to include the PQC scheme [SPHINCS+](https://sphincs.org/), which is about to be NIST [standardized](https://csrc.nist.gov/News/2022/pqc-candidates-to-be-standardized-and-round-4).
However, work has stalled as people were busy with other stuff.

After talking to @lukpueh we decided to directly include SPHINCS+ into the "new" Signer API.

I welcome feedback and additional TODOs.

### CODE REVIEW

File: securesystemslib/formats.py [hunk 1]
```diff
@@ -256,19 +256,30 @@
...
 ED25519PUBLIC_SCHEMA = SCHEMA.LengthBytes(32)
 
+# An SPHINCS raw public key, which must be 32 bytes.
+SPHINCSPUBLIC_SCHEMA = SCHEMA.LengthBytes(32)
+
 # An ED25519 raw seed key, which must be 32 bytes.
 ED25519SEED_SCHEMA = SCHEMA.LengthBytes(32)
 
+# An SPHINCS raw public key, which must be 32 bytes.
+SPHINCSPRIVATE_SCHEMA = SCHEMA.LengthBytes(64)
```
Comment (jku) [inline securesystemslib/formats.py:266]: comment seems incorrect

File: securesystemslib/formats.py [hunk 2]
```diff
@@ -256,19 +256,30 @@
...
+
 # An ED25519 raw signature, which must be 64 bytes.
 ED25519SIGNATURE_SCHEMA = SCHEMA.LengthBytes(64)
 
+# An SPHINCS raw signature, which must be 64 bytes.
+SPHINCSSIGNATURE_SCHEMA = SCHEMA.LengthBytes(7_856)
```
Comment (jku) [inline securesystemslib/formats.py:272]: comment byte length mismatch?

**Review by jku — COMMENTED** [review]
I don't know anything about sphincs so just commenting on cosmetics: there's a couple of code comments that do not seem to match code. I think it would be best to not duplicate information like bytearray length in the comments if the code is clear as it seems to be here

**Comment by rugo** [comment]
> I don't know anything about sphincs so just commenting on cosmetics: there's a couple of code comments that do not seem to match code. I think it would be best to not duplicate information like bytearray length in the comments if the code is clear as it seems to be here

Makes sense. I removed them.

File: securesystemslib/formats.py [hunk 3]
```diff
@@ -193,7 +193,7 @@
 # Supported securesystemslib key types.
 KEYTYPE_SCHEMA = SCHEMA.OneOf(
   [SCHEMA.String('rsa'), SCHEMA.String('ed25519'), SCHEMA.String('ecdsa'),
-   SCHEMA.RegularExpression(r'ecdsa-sha2-nistp(256|384)')])
+   SCHEMA.RegularExpression(r'ecdsa-sha2-nistp(256|384)'), SCHEMA.String('sphincs-shake-128s')])
```
Comment (lukpueh) [inline securesystemslib/formats.py:196]: Does it make sense to call the keytype just `sphincs`? (see discussion about keytype vs. scheme in #239 and #308)

File: securesystemslib/sphincs_keys.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    return shake_128s.verify(data, signature, public_key)
+
+
+if __name__ == '__main__':
+    # The interactive sessions of the documentation strings can
+    # be tested by running 'ed25519_keys.py' as a standalone module.
+    # python -B ed25519_keys.py
+    import doctest
+    doctest.testmod()
```
Comment (lukpueh) [inline securesystemslib/sphincs_keys.py:88]: No need to add `doctest` boilerplate, there are no doctests in this module, and we wouldn't even run them in CI.

**Review by lukpueh — COMMENTED** [review]
Thanks for the patch, @rugo! I think what's there is great. There are a few peculiarities that were likely inspired from the existing interfaces (e.g. passing args that are not used, keytype/scheme redundancy), but I think it's okay for the sake of consistency.

Would you mind adding tests to `check_public_interfaces.TestPublicInterfaces.test_keys`, to assert that we fail gracefully if `PySPX` is not installed?

It would also be good to have `interface.generate_and_write_sphincs*`, and `interface.import_sphincs*` functions as we have for other key types. But that can be added in a separate PR.

**Comment by rugo** [comment]
@lukpueh Thanks for the feedback. I added the required tests, changed the `keytype` and removed `doctest`.

Regadring `interface.generate_and_write_sphincs`: I would prefer to address this in a separate PR. This would need more coordination as SPHINCS+ does not have a standardized key OID yet (AFAIK). Decoding a SPHINCS+ key from PEM would therefore require us to create our own non-standard key identifier and format.

**Comment by rugo** [comment]
@lukpueh Could you check the failing test and give me some more insights? I don't really understand what is happening as the error says:

```
subprocess.TimeoutExpired: Command '['gpg', '--detach-sign', '--digest-algo', 'SHA256', '--local-user', 'E8AC80C924116DABB51D4B987CB07D6D2C199C7C', '--homedir', '/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmpuwgkkmy9/rsa']' timed out after 3 seconds
```

Which doesn't seem related to what I did. Did the `macos` test case break or am I missing something?

**Comment by lukpueh** [comment]
@rugo: I just merged  #439, which auto-formats the entire code base. The easiest thing might be to re-apply your changes manually on top of the updated master branch, run [auto-formatters commands (w/o `--check --diff` options)](https://github.com/secure-systems-lab/securesystemslib/blob/600aca1091b6c4de048d53ae1d923ca98ff4d773/tox.ini#L47-L49) over them, and push them in a single commit to this branch. I apologize for the inconvenience, and am happy to help resolve any conflicts.

**Comment by rugo** [comment]
Alright. I think I have addressed your points. Additional feedback welcome.

**Comment by rugo** [comment]
Since the pylint test case complained about some minor things I fixed them now.

**Comment by rugo** [comment]
I rebased again to avoid the `requirements.txt` merge conflict.

File: securesystemslib/keys.py [hunk 5]
```diff
@@ -349,6 +350,33 @@ def generate_ed25519_key(scheme="ed25519"):
     return ed25519_key
 
 
+def generate_sphincs_key(scheme="sphincs-shake-128s"):
+    """Generate a SPHINCS+ key pair.
+    Arguments:
+          scheme (str): Name of the scheme as defined in formats.py.
+    Returns:
+        dict: A dictionary containing the ED25519 keys.
```
[the remark this answers is not in the document]
Comment (rugo) [reply securesystemslib/keys.py]: :see_no_evil: fixed

**Review by lukpueh — APPROVED** [review]
💯

**Pull request merged**